libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/RouterWrapper.kt

```diff
@@ -0,0 +1,200 @@
+                                    """.trimIndent()
+                                )
+                            )
+
+                            callback.onFailure(failureReasons, routeOptions)
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    override fun getRouteRefresh(
+        route: DirectionsRoute,
+        legIndex: Int,
+        callback: RouteRefreshCallback
+    ): Long {
+        val routeOptions = route.routeOptions()!!
+        val userAndProfile = "${routeOptions.user()}/${routeOptions.profile()}"
+        val refreshOptions = RouteRefreshOptions(
+            route.requestUuid() ?: "",
+            route.routeIndex()?.toIntOrNull() ?: 0,
+            legIndex,
+            RoutingProfile(userAndProfile),
+        )
+
+        return router.getRouteRefresh(refreshOptions, route.toJson()) { result, _ ->
+            mainJobControl.scope.launch {
+                if (result.isValue) {
+                    val refreshedRoute =
+                        withContext(ThreadController.IODispatcher) {
+                            DirectionsRoute.fromJson(
+                                result.value!!,
+                                routeOptions,
+                                route.requestUuid()
+                            )
+                        }
+                    callback.onRefresh(refreshedRoute)
+                } else {
+                    result.error!!.run {
+                        val errorMessage =
+                            """
+                                Route refresh failed.
+                                error = $error
+                                legIndex = $legIndex
+                                code = $code
+                                requestId = $requestId
+                            """.trimIndent()
+
+                        logW(TAG, Message(errorMessage))
+
+                        callback.onError(
+                            RouteRefreshError("Route refresh failed", Exception(errorMessage))
+                        )
+                    }
+                }
+            }
+        }
+    }
+
+    override fun cancelRouteRequest(requestId: Long) {
+        router.cancelRequest(requestId)
+    }
+
+    override fun cancelRouteRefreshRequest(requestId: Long) {
+        router.cancelRequest(requestId)
+    }
+
+    override fun cancelAll() {
+        router.cancelAll()
+    }
+
+    override fun shutdown() {
+        router.cancelAll()
+    }
+
+    private suspend fun parseDirectionsResponse(
+        json: String,
+        options: RouteOptions?
+    ): List<DirectionsRoute> =
+        withContext(ThreadController.IODispatcher) {
+            val jsonObject = JSONObject(json)
+            val uuid: String? = if (jsonObject.has(UUID)) {
+                jsonObject.getString(UUID)
+            } else {
+                null
+            }
+
+            val response = DirectionsResponse.fromJson(json, options, uuid)
```

Comment:
Just a note that this could be automated and exposed as `DirectionsResponse.fromJson(json, options)` - https://github.com/mapbox/mapbox-java/issues/1292.

Reply:
ok, will be simplified when mapbox-java will have it

Reply:
added TODO